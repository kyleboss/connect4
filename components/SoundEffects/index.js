export * from './SoundEffects';
export { default } from './SoundEffects';
