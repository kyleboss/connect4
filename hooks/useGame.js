import { useContext } from "react";

import GameContext from "@/state/GameContext";

const useGame = () => useContext(GameContext);

export default useGame;
