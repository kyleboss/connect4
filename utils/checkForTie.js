const checkForTie = (board) =>
  board.every((row) => row.every((cell) => cell !== null));

module.exports = checkForTie;
