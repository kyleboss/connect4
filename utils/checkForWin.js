const checkHorizontal = (player, board) => {
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row][col + 1] === player &&
        board[row][col + 2] === player &&
        board[row][col + 3] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkVertical = (player, board) => {
  for (let col = 0; col < 7; col++) {
    for (let row = 0; row < board.length - 3; row++) {
      if (
        board[row][col] === player &&
        board[row + 1][col] === player &&
        board[row + 2][col] === player &&
        board[row + 3][col] === player
      ) {
        return true;
      }
    }
  }
  return false;
};

const checkDiagonal = (player, board) => {
  // Check right diagonal
  for (let row = 0; row < board.length - 3; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row + 1][col + 1] === player &&
        board[row + 2][col + 2] === player &&
        board[row + 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  // Check left diagonal
  for (let row = 3; row < board.length; row++) {
    for (let col = 0; col < board[row].length - 3; col++) {
      if (
        board[row][col] === player &&
        board[row - 1][col + 1] === player &&
        board[row - 2][col + 2] === player &&
        board[row - 3][col + 3] === player
      ) {
        return true;
      }
    }
  }

  return false;
};

const checkForWin = (player, board) =>
  checkHorizontal(player, board) ||
  checkVertical(player, board) ||
  checkDiagonal(player, board);

// Must be CommonJS to accomodate for express server.
module.exports = checkForWin;
