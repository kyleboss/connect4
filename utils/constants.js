// Game State
const GAME_STATE_WAITING = "GAME_STATE_WAITING";
const GAME_STATE_ACTIVE = "GAME_STATE_ACTIVE";
const GAME_STATE_GAME_OVER = "GAME_STATE_GAME_OVER";

// User roles
const USER_ROLE_GUEST = "USER_ROLE_GUEST";
const USER_ROLE_RED = "USER_ROLE_RED";
const USER_ROLE_YELLOW = "USER_ROLE_YELLOW";
const USER_ROLE_NONE = "USER_ROLE_NONE";
const USER_ROLES = {
  [USER_ROLE_GUEST]: "Guest",
  [USER_ROLE_RED]: "Red",
  [USER_ROLE_YELLOW]: "Yellow",
  [USER_ROLE_NONE]: "No one",
};

// URLs
const CLIENT_PORT = 3000;
const SERVER_PORT = 3001;
const CLIENT_URL = `http://localhost:${CLIENT_PORT}`;
const SERVER_URL = `http://localhost:${SERVER_PORT}`;

// Socket Events
const SOCKET_EVENT_JOIN_GAME = "SOCKET_EVENT_JOIN_GAME";
const SOCKET_EVENT_RESET_GAME = "SOCKET_EVENT_RESET_GAME";
const SOCKET_EVENT_MAKE_MOVE = "SOCKET_EVENT_MAKE_MOVE";
const SOCKET_EVENT_ASSIGN_ROLE = "SOCKET_EVENT_ASSIGN_ROLE";
const SOCKET_EVENT_UPDATE_STATE = "SOCKET_EVENT_UPDATE_STATE";
const SOCKET_EVENT_CONNECTION = "connection"; // Official socket event.
const SOCKET_EVENT_DISCONNECT = "disconnect"; // Official socket event.

// Must be CommonJS to accomodate for express server.
module.exports = {
  GAME_STATE_WAITING,
  GAME_STATE_ACTIVE,
  GAME_STATE_GAME_OVER,
  USER_ROLE_GUEST,
  USER_ROLE_RED,
  USER_ROLE_YELLOW,
  USER_ROLE_NONE,
  USER_ROLES,
  CLIENT_PORT,
  SERVER_PORT,
  CLIENT_URL,
  SERVER_URL,
  SOCKET_EVENT_JOIN_GAME,
  SOCKET_EVENT_RESET_GAME,
  SOCKET_EVENT_MAKE_MOVE,
  SOCKET_EVENT_ASSIGN_ROLE,
  SOCKET_EVENT_UPDATE_STATE,
  SOCKET_EVENT_CONNECTION,
  SOCKET_EVENT_DISCONNECT,
};
