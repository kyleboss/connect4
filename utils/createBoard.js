const createBoard = () => Array(6).fill(Array(7).fill(null));

module.exports = createBoard;
