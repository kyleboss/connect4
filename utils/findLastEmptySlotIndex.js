const findLastEmptySlotIndex = (board, columnIndex) => {
  for (let i = board.length - 1; i >= 0; i--) {
    if (!board[i][columnIndex]) {
      return i;
    }
  }
  return -1; // Indicates the column is full
};

module.exports = findLastEmptySlotIndex;
