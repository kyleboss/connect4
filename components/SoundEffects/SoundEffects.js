import { useCallback, useEffect, useState, useRef } from "react";
import useSound from "use-sound";
import styled, { css, keyframes } from "styled-components";

import notificationSound from "@/sounds/notification.mp3";
import winnerSound from "@/sounds/winner.mp3";
import useGame from "@/hooks/useGame";

const shakeAnimation = keyframes`
  0% { transform: rotate(0deg); }
  20% { transform: rotate(-10deg); }
  40% { transform: rotate(10deg); }
  60% { transform: rotate(-5deg); }
  80% { transform: rotate(5deg); }
  100% { transform: rotate(0deg); }
`;

const Container = styled.span.attrs({ role: "button" })`
  font-size: 3rem;
  position: fixed;
  bottom: 20px;
  left: 20px;
  transition: transform 0.3s ease;
  cursor: pointer;

  &:hover {
    animation: ${css`
      ${shakeAnimation} 0.5s ease-in-out
    `};
  }
`;

const SoundEffects = () => {
  const [soundEnabled, setSoundEnabled] = useState(true);
  const [playWinnerSound] = useSound(winnerSound);
  const [playNotificationSound] = useSound(notificationSound);
  const { currentPlayer, userRole, winner } = useGame();
  const justEnabledSound = useRef(false);

  useEffect(() => {
    if (
      soundEnabled &&
      currentPlayer === userRole &&
      !justEnabledSound.current
    ) {
      playNotificationSound();
    }

    justEnabledSound.current = false;
  }, [currentPlayer, playNotificationSound, soundEnabled, userRole]);

  useEffect(() => {
    if (soundEnabled && winner === userRole) {
      playWinnerSound();
    }
  }, [currentPlayer, playWinnerSound, soundEnabled, userRole, winner]);

  const handleClick = useCallback(() => {
    setSoundEnabled((prevSoundEnabled) => {
      if (!prevSoundEnabled) {
        justEnabledSound.current = true;
      }
      return !prevSoundEnabled;
    });
  }, []);

  return (
    <Container onClick={handleClick}>{soundEnabled ? "ğŸ””" : "ğŸ”•"}</Container>
  );
};

export default SoundEffects;
