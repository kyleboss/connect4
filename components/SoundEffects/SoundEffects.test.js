import React from "react";
import { render, fireEvent } from "@testing-library/react";
import useSound from "use-sound";

import SoundEffects from "./SoundEffects";
import useGame from "../../hooks/useGame";
import { USER_ROLE_RED } from "../../utils/constants";

jest.mock("use-sound", () =>
  jest.fn().mockImplementation((sound) => [jest.fn(), { sound }])
);

jest.mock("../../hooks/useGame", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    currentPlayer: USER_ROLE_RED,
    userRole: USER_ROLE_RED,
    winner: USER_ROLE_RED,
  })),
}));

describe("SoundEffects", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<SoundEffects />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("plays notification sound when it is the user's turn", () => {
    const mockPlayNotificationSound = jest.fn();
    useSound.mockImplementation(() => [mockPlayNotificationSound]);
    useGame.mockImplementation(() => ({
      currentPlayer: USER_ROLE_RED,
      userRole: USER_ROLE_RED,
      winner: null,
    }));

    render(<SoundEffects />);
    expect(mockPlayNotificationSound).toHaveBeenCalled();
  });

  it("plays winner sound when the user wins", () => {
    const mockPlayWinnerSound = jest.fn();
    useSound.mockImplementation(() => [mockPlayWinnerSound]);
    useGame.mockImplementation(() => ({
      currentPlayer: USER_ROLE_RED,
      userRole: USER_ROLE_RED,
      winner: USER_ROLE_RED,
    }));

    render(<SoundEffects />);
    expect(mockPlayWinnerSound).toHaveBeenCalled();
  });

  it("toggles sound on click", () => {
    const { getByRole } = render(<SoundEffects />);
    const soundButton = getByRole("button");
    fireEvent.click(soundButton);
    expect(soundButton).toHaveTextContent("ðŸ”•");
    fireEvent.click(soundButton);
    expect(soundButton).toHaveTextContent("ðŸ””");
  });
});
