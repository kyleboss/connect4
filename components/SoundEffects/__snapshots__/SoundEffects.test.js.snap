// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SoundEffects renders correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-7431bd22-0 hhzhNH"
    role="button"
  >
    ðŸ””
  </span>
</DocumentFragment>
`;
